internal/kallsyms/kallsyms.go

```diff
@@ -72,3 +74,56 @@ func loadKernelModuleMapping(f io.Reader) (map[string]string, error) {
 	}
 	return mods, nil
 }
+
+func LoadSymbolAddresses(symbols []string) (map[string]uint64, error) {
+	if len(symbols) == 0 {
+		return map[string]uint64{}, nil
+	}
+
+	f, err := os.Open("/proc/kallsyms")
+	if err != nil {
+		return nil, err
+	}
+
+	ksyms, err := loadSymbolAddresses(f, symbols)
+	if err != nil {
+		return nil, err
+	}
+
+	return ksyms, nil
+}
+
+func loadSymbolAddresses(f io.Reader, symbols []string) (map[string]uint64, error) {
```

Comment:
Is this faster than using a map? e.g.
`func loadSymbolAddresses(f io.Reader, syms map[string]uintptr) error`, looking up every symbol read from kallsyms against syms and setting the value if it's present.

Reply:
> Is this faster than using a map?

Technically, using a map is faster, especially for large amounts of symbols. But if you go an pprof this, then 99% of time is spent allocating in the scanner.Text and the read syscall. So we can essentially go whatever way we find easier to use, doesn't make a practical difference.

Reply:
I did some quick benchmarks and on my machine the slice is faster (input of 19 elements, where one of the requested symbols is the last in `/proc/kallsyms`):
```
BenchmarkLoadSymbolAddresses/map_output-4                      7         162487483 ns/op
BenchmarkLoadSymbolAddresses/in_out-4                          7         165968776 ns/op
```

Here is the benchmark code:
This is the new function
```go
func loadSymbolAddressesInOut(f io.Reader, symbols map[string]uint64) error {
	result := make(map[string]uint64)

	scan := bufio.NewScanner(f)
	for scan.Scan() {
		var (
			addr   uint64
			t      rune
			symbol string
		)

		line := scan.Text()

		_, err := fmt.Sscanf(line, "%x %c %s", &addr, &t, &symbol)
		if err != nil {
			return err
		}
		if _, found := symbols[symbol]; found {
			result[symbol] = addr
		}
		// We got all the symbols we need, no need to process any more lines.
		if len(result) == len(symbols) {
			break
		}
	}

	if scan.Err() != nil {
		return scan.Err()
	}

	return nil
}
```
This are the benchmarks:
```go
var globalMap map[string]uint64

func BenchmarkLoadSymbolAddresses(b *testing.B) {
	kallsyms, err := os.ReadFile("/proc/kallsyms")
	if err != nil {
		b.Fatal(err)
	}

	b.ResetTimer()

	b.Run("map output", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			krdr := bytes.NewBuffer(kallsyms)
			result, err := loadSymbolAddresses(krdr, []string{
				"bpf_fd_inode_storage_update_elem",
				"kprobe_insn_page",
				"el1t_64_sync",
				"el1t_64_irq",
				"el1t_64_fiq",
				"el1t_64_error",
				"el1h_64_sync",
				"el1h_64_irq",
				"el1h_64_fiq",
				"el1h_64_error",
				"el0t_64_sync",
				"el0t_64_irq",
				"el0t_64_fiq",
				"el0t_64_error",
				"el0t_32_sync",
				"el0t_32_irq",
				"el0t_32_fiq",
				"el0t_32_error",
			})
			if err != nil {
				b.Fatal(err)
			}
			globalMap = result
		}
	})

	b.Run("in out", func(b *testing.B) {
		for i := 0; i < b.N; i++ {
			krdr := bytes.NewBuffer(kallsyms)
			result := map[string]uint64{
				"bpf_fd_inode_storage_update_elem": 0,
				"kprobe_insn_page":                 0,
				"el1t_64_sync":                     0,
				"el1t_64_irq":                      0,
				"el1t_64_fiq":                      0,
				"el1t_64_error":                    0,
				"el1h_64_sync":                     0,
				"el1h_64_irq":                      0,
				"el1h_64_fiq":                      0,
				"el1h_64_error":                    0,
				"el0t_64_sync":                     0,
				"el0t_64_irq":                      0,
				"el0t_64_fiq":                      0,
				"el0t_64_error":                    0,
				"el0t_32_sync":                     0,
				"el0t_32_irq":                      0,
				"el0t_32_fiq":                      0,
				"el0t_32_error":                    0,
			}
			err := loadSymbolAddressesInOut(krdr, result)
			if err != nil {
				b.Fatal(err)
			}
			globalMap = result
		}
	})

}
```